AX: AXIsolatedTree::firstMarker should be constructed from the web area object

hoffmanjoshua opened the pull request:
#### b8a516d7b28caa8059fccb036c8df935b4dcb75b
<pre>
AX: AXIsolatedTree::firstMarker should be constructed from the web area object
<a href="https://bugs.webkit.org/show_bug.cgi?id=280488">https://bugs.webkit.org/show_bug.cgi?id=280488</a>
<a href="https://rdar.apple.com/136793644">rdar://136793644</a>

Reviewed by Tyler Wilcock and Andres Gonzalez.

When we are mixing AX-thread text markers and main thread text markers, we
need to convert AXTextMarkers to CharacterPositions. CharacterPositions
are based on nodes, so using the root object (the ScrollView) will result
in a null CharacterPosition. Instead, let&apos;s use the WebArea which has an
associated node.

Ran against existing ax-thread-text-api/ tests that exercise startMarker.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::webArea):
(WebCore::AXIsolatedTree::firstMarker):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: <a href="https://commits.webkit.org/284373@main">https://commits.webkit.org/284373@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fb64bdf952cb46b244697341cffc54bba343522f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69170 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48570 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21842 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73251 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20328 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71287 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56371 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20177 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55051 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13501 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72236 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44334 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59717 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/35530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17147 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18702 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62946 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17492 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74962 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13152 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16730 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13190 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59800 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62607 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15359 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10609 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4216 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44374 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45447 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46643 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -987,6 +987,17 @@ RefPtr<AXIsolatedObject> AXIsolatedTree::rootNode()
     return m_rootNode;
 }
 
+RefPtr<AXIsolatedObject> AXIsolatedTree::webArea()
```
Maybe `rootWebArea()` instead of `webArea()`? Since technically there can be more than one in a non-site-isolated WebKit when `iframes` are on the page.

hoffmanjoshua replied:
Good point, then it will match what we have in `AXObjectCache` too

twilco commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1352,8 +1363,12 @@ void AXIsolatedTree::processQueuedNodeUpdates()
-    RefPtr root = rootNode();
-    return root ? AXTextMarker { root->treeID(), root->objectID(), 0 } : AXTextMarker();
+    ASSERT(!isMainThread());
+    // The first marker should be constructed from the WebArea, not the ScrollView (the root),
```
Maybe "The first marker should be constructed from the root WebArea, not the true root of the tree which is the ScrollView, so that..."

Take it or leave it, just a bit more precise that we are looking specifically for the root web area.

hoffmanjoshua replied:
Updated!

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -987,6 +987,17 @@ RefPtr<AXIsolatedObject> AXIsolatedTree::rootNode()
 
+RefPtr<AXIsolatedObject> AXIsolatedTree::rootWebArea()
+{
+    AXTRACE("AXIsolatedTree::webArea"_s);
```
In AXTRACE we pass the exact name of the function: webArea -> rootWebArea

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -987,6 +987,17 @@ RefPtr<AXIsolatedObject> AXIsolatedTree::rootNode()
+    ASSERT(!isMainThread());
+
+    RefPtr root = rootNode();
+    return Accessibility::findChild(*root, [] (auto& object) {
```
Should we check for !root?

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1352,8 +1363,12 @@ void AXIsolatedTree::processQueuedNodeUpdates()
+    // The first marker should be constructed from the root WebArea, not the true root of the tree
+    // which is the ScrollView, so that when we convert the marker to a CharacterPosition, there
+    // is an associated node. Otherwise, the CharacterPosition will be null.
+    RefPtr rootWebArea = webArea();
```
Thought that the method is called rootWebArea.

hoffmanjoshua replied:
Updated, thanks for catching this!

AndresGonzalezApple commented on Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
```diff
@@ -1352,8 +1363,12 @@ void AXIsolatedTree::processQueuedNodeUpdates()
+    // which is the ScrollView, so that when we convert the marker to a CharacterPosition, there
+    // is an associated node. Otherwise, the CharacterPosition will be null.
+    RefPtr rootWebArea = webArea();
+    return rootWebArea ? AXTextMarker { rootWebArea->treeID(), rootWebArea->objectID(), 0 } : AXTextMarker();
```
Can we write:
return rootWebArea ? { rootWebArea->treeID(), rootWebArea->objectID(), 0 } : AXTextMarker();
Maybe not, if the compiler cannot infer the type within the ? : statement.

hoffmanjoshua replied:
I tried that, but unfortunately the compiler cannot infer the type here.

hoffmanjoshua committed:
AX: AXIsolatedTree::firstMarker should be constructed from the web area object
https://bugs.webkit.org/show_bug.cgi?id=280488
rdar://136793644

Reviewed by Tyler Wilcock and Andres Gonzalez.

When we are mixing AX-thread text markers and main thread text markers, we
need to convert AXTextMarkers to CharacterPositions. CharacterPositions
are based on nodes, so using the root object (the ScrollView) will result
in a null CharacterPosition. Instead, let's use the WebArea which has an
associated node.

Ran against existing ax-thread-text-api/ tests that exercise startMarker.

* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::AXIsolatedTree::webArea):
(WebCore::AXIsolatedTree::firstMarker):
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.h:

Canonical link: https://commits.webkit.org/284373@main
